src/protocols/secure_channel/CASESession.cpp

```diff
@@ -404,187 +445,283 @@ CHIP_ERROR CASESession::HandleSigmaR1(System::PacketBufferHandle && msg)
-        SuccessOrExit(err = tlvWriter.PutBytes(TLV::ContextTag(4), msg_R2_Encrypted.Get(),
-                                               static_cast<uint32_t>(msg_r2_signed_enc_len + kTAGSize)));
-        SuccessOrExit(err = tlvWriter.EndContainer(outerContainerType));
-        SuccessOrExit(err = tlvWriter.Finalize(&msg_R2));
-    }
+    tlvWriterMsg2.Init(std::move(msg_R2));
+    ReturnErrorOnFailure(tlvWriterMsg2.StartContainer(TLV::AnonymousTag, TLV::kTLVType_Structure, outerContainerType));
+    ReturnErrorOnFailure(tlvWriterMsg2.PutBytes(TLV::ContextTag(1), &msg_rand[0], sizeof(msg_rand)));
+    ReturnErrorOnFailure(tlvWriterMsg2.Put(TLV::ContextTag(2), GetLocalSessionId(), true));
+    ReturnErrorOnFailure(
+        tlvWriterMsg2.PutBytes(TLV::ContextTag(3), mEphemeralKey.Pubkey(), static_cast<uint32_t>(mEphemeralKey.Pubkey().Length())));
+    ReturnErrorOnFailure(tlvWriterMsg2.PutBytes(TLV::ContextTag(4), msg_R2_Encrypted.Get(),
+                                                static_cast<uint32_t>(msg_r2_signed_enc_len + kTAGSize)));
+    ReturnErrorOnFailure(tlvWriterMsg2.EndContainer(outerContainerType));
+    ReturnErrorOnFailure(tlvWriterMsg2.Finalize(&msg_R2));
 
-    err = mCommissioningHash.AddData(ByteSpan{ msg_R2->Start(), msg_R2->DataLength() });
-    SuccessOrExit(err);
+    ReturnErrorOnFailure(mCommissioningHash.AddData(ByteSpan{ msg_R2->Start(), msg_R2->DataLength() }));
 
-    mNextExpectedMsg = Protocols::SecureChannel::MsgType::CASE_SigmaR3;
+    mState = kSentSigma2;
 
     // Call delegate to send the msg to peer
-    err = mExchangeCtxt->SendMessage(Protocols::SecureChannel::MsgType::CASE_SigmaR2, std::move(msg_R2),
-                                     SendFlags(SendMessageFlags::kExpectResponse));
-    SuccessOrExit(err);
+    ReturnErrorOnFailure(mExchangeCtxt->SendMessage(Protocols::SecureChannel::MsgType::CASE_Sigma2, std::move(msg_R2),
+                                                    SendFlags(SendMessageFlags::kExpectResponse)));
 
-    ChipLogDetail(SecureChannel, "Sent SigmaR2 msg");
+    ChipLogDetail(SecureChannel, "Sent Sigma2 msg");
 
-exit:
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR CASESession::HandleSigma2Resume(System::PacketBufferHandle && msg)
+{
+    CHIP_ERROR err = CHIP_NO_ERROR;
+    System::PacketBufferTLVReader tlvReader;
+    TLV::TLVType containerType = TLV::kTLVType_Structure;
+
+    uint16_t responderSessionId;
+
+    uint32_t decodeTagIdSeq = 0;
+
+    ChipLogDetail(SecureChannel, "Received Sigma2Resume msg");
+
+    uint8_t sigma2ResumeMIC[kTAGSize];
+
+    tlvReader.Init(std::move(msg));
+    SuccessOrExit(err = tlvReader.Next(containerType, TLV::AnonymousTag));
+    SuccessOrExit(err = tlvReader.EnterContainer(containerType));
+
+    SuccessOrExit(err = tlvReader.Next());
+    VerifyOrExit(TLV::TagNumFromTag(tlvReader.GetTag()) == ++decodeTagIdSeq, err = CHIP_ERROR_INVALID_TLV_TAG);
+    SuccessOrExit(err = tlvReader.GetBytes(mResumptionId, kCASEResumptionIDSize));
```

Comment:
Should validate that `GetLength()` returns *precisely* == kCASEResumptionIDSize. GetBytes() currently expects a buffer size >= to the tag length. If the tag is too small, the resumption ID will be partially filled.